### ISSUE TITLE
Inconsistent behavior of time zone conversion

### ISSUE THREAD

**Issue by wangguansong** [r-dbi/RMariaDB#80]
This has puzzled me for a while.

The database is set into UTC (GMT) +8 time zone (China). I have no problem in MySQL Workbench, while the date returned from RMariaDB is not what I expected sometimes.

Even though the session time zone is supposed to be set to UTC since version 0.11-1 (2016-03-24), the `from_unixtime` function of MySQL is still using +8 time zone, however `date` seems to be using UTC time zone.

At the mean time, RMySQL is returning things as I expected.

Here are some testing:

```r
> packageVersion("RMariaDB")
[1] ‘1.0.2’
> packageVersion("RMySQL")
[1] ‘0.10.13’
> Sys.timezone()
[1] "Asia/Shanghai"
> as.POSIXct(86400-3600, origin = "1970-01-01 00:00:00")
[1] "1970-01-02 07:00:00 CST"

> library("RMariaDB")
> dbcon <- RMariaDB::dbConnect(...)
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              +00:00
> dbGetQuery(dbcon, "select from_unixtime(86400-3600)")
  from_unixtime(86400-3600)
1       1970-01-02 07:00:00
> dbGetQuery(dbcon,"select date(from_unixtime(86400-3600))")
  date(from_unixtime(86400-3600))
1                      1970-01-01
> ###### Expected to be "1970-01-02" ######
> dbGetQuery(dbcon, "select date_format(from_unixtime(86400-3600), '%Y-%m-%d')")
  date_format(from_unixtime(86400-3600), '%Y-%m-%d')
1                                         1970-01-01
> dbExecute(dbcon, 'SET time_zone = "+08:00"')
[1] 0
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              +08:00
> dbGetQuery(dbcon, "select from_unixtime(86400-3600)")
  from_unixtime(86400-3600)
1       1970-01-02 15:00:00
> ###### Expected to be "1970-01-02 07:00:00" ######
> dbGetQuery(dbcon,"select date(from_unixtime(86400-3600))")
  date(from_unixtime(86400-3600))
1                      1970-01-02
> dbGetQuery(dbcon, "select date_format(from_unixtime(86400-3600), '%Y-%m-%d')")
  date_format(from_unixtime(86400-3600), '%Y-%m-%d')
1                                         1970-01-02
> dbDisconnect(dbcon)
> as.POSIXct(86400-3600, origin = "1970-01-01 00:00:00")
[1] "1970-01-02 07:00:00 CST"

> library("RMySQL")
> dbcon <- RMySQL::dbConnect(...)
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              SYSTEM
> Sys.timezone()
[1] "Asia/Shanghai"
> dbGetQuery(dbcon, "select from_unixtime(86400-3600)")
  from_unixtime(86400-3600)
1       1970-01-02 07:00:00
> dbGetQuery(dbcon,"select date(from_unixtime(86400-3600))")
  date(from_unixtime(86400-3600))
1                      1970-01-02
> dbGetQuery(dbcon, "select date_format(from_unixtime(86400-3600), '%Y-%m-%d')")
  date_format(from_unixtime(86400-3600), '%Y-%m-%d')
1                                         1970-01-02
> dbExecute(dbcon, 'SET time_zone = "+08:00"')
[1] 0
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              +08:00
> dbGetQuery(dbcon, "select from_unixtime(86400-3600)")
  from_unixtime(86400-3600)
1       1970-01-02 07:00:00
> dbGetQuery(dbcon,"select date(from_unixtime(86400-3600))")
  date(from_unixtime(86400-3600))
1                      1970-01-02
> dbGetQuery(dbcon, "select date_format(from_unixtime(86400-3600), '%Y-%m-%d')")
  date_format(from_unixtime(86400-3600), '%Y-%m-%d')
1                                         1970-01-02
> dbExecute(dbcon, 'SET time_zone = "+00:00"')
[1] 0
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              +00:00
> dbGetQuery(dbcon, "select from_unixtime(86400-3600)")
  from_unixtime(86400-3600)
1       1970-01-01 23:00:00
> dbGetQuery(dbcon,"select date(from_unixtime(86400-3600))")
  date(from_unixtime(86400-3600))
1                      1970-01-01
> dbGetQuery(dbcon, "select date_format(from_unixtime(86400-3600), '%Y-%m-%d')")
  date_format(from_unixtime(86400-3600), '%Y-%m-%d')
1                                         1970-01-01
> dbDisconnect(dbcon)
[1] TRUE
```

**Comment by krlmlr** [comment]
Thanks. RMariaDB doesn't set the `"tzone"` attribute for the date-times values it returns. This means that R will display times in your local time zone by default. Date values don't seem to have such an attribute:

``` r
library(DBI)
con <- dbConnect(RMariaDB::MariaDB())
zero_time <- dbGetQuery(con, "SELECT from_unixtime(0)")[[1]]
zero_time
#> [1] "1970-01-01 01:00:00 CET"
lubridate::with_tz(zero_time, "UTC")
#> [1] "1970-01-01 UTC"
```

Created on 2017-12-21 by the [reprex package](http://reprex.tidyverse.org) (v0.1.1.9000).

The time zone translations requires setting the `time_zone` database variable, which is done when connecting. If you're changing that database variable, you're responsible of manually assigning a `"tzone"` attribute. I'd advise against this, and translate time zones in R if necessary.

**Comment by wangguansong** [comment]
> RMariaDB doesn't set the "tzone" attribute for the date-times values it returns.

Is this intended? I think this is the key to my issue. If I change the connection session `time_zone`, the returning date-time in that `time_zone` is still treated as UTC time in R. As a result, the original time stamp (`0`) is converted twice after the query.

```r
> dbExecute(dbcon, 'SET time_zone = "+08:00"')
[1] 0
> dbGetQuery(dbcon, "SELECT @@global.time_zone, @@session.time_zone")
  @@global.time_zone @@session.time_zone
1             SYSTEM              +08:00
> dbGetQuery(dbcon, "select from_unixtime(0)")[[1]]
[1] "1970-01-01 16:00:00 CST"
> dbGetQuery(dbcon, "select convert(from_unixtime(0), char(20))")[[1]]
[1] "1970-01-01 08:00:00"
```

It would be nice if RMariaDB can set the `tzone` attribute of the `POSIXct` class with the session `time_zone`.

The reason I want to set the `time_zone` variable is that the statistics grouped by date will be messed up under UTC since they are recorded in local time zone. I'm writing a wrapper data query package for a team and I'd like to keep them from handling time zone issues.

If this is as intended, I guess I can set the `time_zone` to `+8` and then adjust `POSIXct` columns if there is any before returning the query.

**Comment by krlmlr** [comment]
Please do not set the `time_zone` in RMariaDB, the default should just work. Are you sure the times are stored in the correct time zone in your database?

**Comment by krlmlr** [comment]
The documentation reads that times are always stored as UTC internally: https://mariadb.com/kb/en/library/time-zones/

**Comment by wangguansong** [comment]
I know little about database management. While as I see it, the actual time stamp written in DB is always UTC. The database system has a `time_zone` variable, which will be in use when I call `from_unixtime`.

RMariaDB set the `time_zone` of the connection session to UTC, so when I group something by date, it's actually by a UTC day but not my local day. To correct this I have to do this `dbExecute(dbcon, 'SET time_zone = "+08:00"')` before the "group by date" query.

**Comment by krlmlr** [comment]
I see. Would `convert_tz()` be an option for you?

``` r
library(DBI)
con <- dbConnect(RMariaDB::MariaDB())
query <-
  "SELECT date(convert_tz(from_unixtime(86400 - 3600), @@session.time_zone, '+08:00'))"
dbGetQuery(con, query)[[1]]
#> [1] "1970-01-02"
```

Created on 2017-12-21 by the [reprex package](http://reprex.tidyverse.org) (v0.1.1.9000).

**Comment by wangguansong** [comment]
For me, yes it would. However, I'm wrapping a `dbGetQuery` for a team and they certainly expect to see the same result as from Navicat or Workbench with the same sql query. It's less ideal to have them make a  `convert_tz` every time they call `date`.

I don't like it either, but I guess to execute `SET time_zone = "+08:00"` then to re-adjust `POSIXct` columns should make it work.

Thanks for the input!

**Comment by krlmlr** [comment]
Thanks. After looking again through the RMariaDB API, it seems that the server always communicates timestamp results according to the `time_zone` variable back to the client, see [source](https://github.com/MariaDB/mariadb-connector-c/blob/3e164b5711135a618cd090f9d3a029ecebe59ef4/include/mysql.h#L382-L395). I'd argue this is a design mistake, but at the same time this will probably stay that way for some time.

Your safest bet might be to change the server's default time zone to UTC (see this [SO answer](https://stackoverflow.com/a/19075291/946850)), but I realize that this might break other applications. RMariaDB could also [query the server's time zone](https://stackoverflow.com/q/2934258/946850) instead of setting the offset to zero, but this is likely to break if the server's time zone is affected by DST.

I see another option: Add a `tzone` argument to `dbConnect()` that, instead of setting the connection's time zone to UTC, makes sure this value is set as `tzone` attribute to all timestamp values returned. Would you like to submit a pull request?

Either way: May I suggest that you teach your users that aggregating by `date()` is brittle in MySQL if used without `convert_tz()`, because you never know what the connection's current time zone settings are?

Obligatory xkcd reference: https://www.xkcd.com/1883/.

**Comment by wangguansong** [comment]
Thanks for the suggestions.

I've never done a pull request before, I'll try and see what I can do about that.

**Comment by krlmlr** [comment]
@wangguansong: Does this problem still affect you?

**Comment by wangguansong** [comment]
> I don't like it either, but I guess to execute SET time_zone = "+08:00" then to re-adjust POSIXct columns should make it work. 

I did a workaround like this and (kinda) solved the issue for me.

**Comment by krlmlr** [comment]
Thanks. Then a `tzone` argument seems like the way to go?

**Comment by jtelleria** [comment]
This is also a big issue for me, as when I call a stored procedure from R using RMariaDB, which does all datetime work in MariaDB itself (Without moving results back to R), and accessing directly the generated results from the MariaDB IDE itself, I can observe than the generated results are incorrect.

This might be due to MariaDB Connector C# itself, as when I execute the same procedures using an ODBC connection from MariaDB IDE I obtain the correct results.

Best,
Juan

**Comment by jtelleriar** [comment]
So the solution to go for me would be that RMariaDB returns the same result than "odbc" package with "MySQL ODBC" or "MariaDB ODBC" Connector Installed (Exactly the same) when dealing with TIMESTAMPs. 

Which can be obtained by:

*By default: SET time_zone = SYSTEM

*When setting a tzone argument in RMariaDB::dbConnect(), include: SET time_zone = '+00:00'

**Comment by SymbolixAU** [comment]
I've forked this repo and added my own `tzone` argument, but it's not having any impact for me; the POSIX time doesn't change.

`tzone` argument added [here](https://github.com/SymbolixAU/RMariaDB/blob/master/R/connect.R#L93)

And here's an example of it running

```
CREATE TABLE `tbl_time` (
	posix_datetime DATETIME,
	posix_varchar VARCHAR(25)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

INSERT INTO tbl_time (posix_datetime, posix_varchar) VALUE ('2018-01-01 00:00:00', '2018-01-01 00:00:00');
```

I've added a `print` statement into the `dbConnect` function so it will print the `SET time_zone` query it's executing

```
library(RMariaDB)

con <- RMariaDB::dbConnect(
	drv = RMariaDB::MariaDB()
	, user = Sys.getenv("USER")
	, host = myHost
	, port = myPort
	, password = Sys.getenv("SQL_PSWD")
	, dbname = myDb
	, tzone = "+10:00"
)
# [1] "SET time_zone = '+10:00'"

dt <- RMariaDB::dbGetQuery(con, "SELECT * FROM tbl_time")
RMariaDB::dbDisconnect(con)
dt
#        posix_datetime       posix_varchar
# 1 2018-01-01 11:00:00 2018-01-01 00:00:00

con <- RMariaDB::dbConnect(
	drv = RMariaDB::MariaDB()
	, user = Sys.getenv("USER")
	, host = myHost
	, port = myPort
	, password = Sys.getenv("SQL_PSWD")
	, dbname = myDb
	, tzone = "-10:00"
)
# [1] "SET time_zone = '-10:00'"

dt <- RMariaDB::dbGetQuery(con, "SELECT * FROM tbl_time")
RMariaDB::dbDisconnect(con)
dt
#        posix_datetime       posix_varchar
# 1 2018-01-01 11:00:00 2018-01-01 00:00:00
```

Both results are the same for a `+10:00` and a `-10:00` time zone

**Comment by jtelleriar** [comment]
I think the issue is with TIMESTAMP DEFAULT in CREATE TABLE, so if no
defaults are used, results shall be the same.

I would always go for SYSTEM time zone in MariaDB, as said before.

Juan

>

**Comment by jtelleria** [comment]
@SymbolixAU @krlmlr As regards the timezone Issue:

If you execute the following statements from an IDE or a MariaDB ODBC Connector MariaDB's @@SESSION.TIME_ZONE (Per Connection) is set as follows by default:
```sql
SELECT @@GLOBAL.TIME_ZONE; -- Result: SYSTEM
SELECT @@SESSION.TIME_ZONE; -- Result: SYSTEM
SELECT @@TIME_ZONE; -- Result: SYSTEM
```
But if you establish a connection through RMariaDB Library, the connection is set as:
```r
con <- dbConnect(
   drv = RMariaDB::MariaDB(), 
   user = "root", 
   password = "*********", 
   host = "localhost", 
   port = 3305, 
   dbname = "eo_mariadb"
)

dbGetQuery(
   con,
   "SELECT @@GLOBAL.TIME_ZONE;"
)

@@GLOBAL.TIME_ZONE
SYSTEM

dbGetQuery(
   con,
   "SELECT @@SESSION.TIME_ZONE;"
)

@@SESSION.TIME_ZONE
+00:00

dbGetQuery(
   con,
   "SELECT @@TIME_ZONE;"
)

@@TIME_ZONE
+00:00
```
We can observe that MariaDB changes the timezone to +00:00, which originates incorrect timestamp defaults in some cases.

The solution is simple: Just as odbc::dbConnect has a timezone argument, RMariaDB::dbConnect could also have one, which shall default to "SET SESSION TIME_ZONE = SYSTEM".

Indeed, if you execute as soon as you establish a connection the following statement, the timestamp issues disappear:
```r
dbExecute(
   con,
   "SET SESSION TIME_ZONE = SYSTEM;"
)
```
I can you if you agree a Pull Request, similar to the one @SymbolixAU proposed but by making use of "SYSTEM" instead of "+00:00" as default. Would you agree? 

Thank you.
Juan

**Comment by jtelleriar** [comment]
@krlmlr I've just done the validation of the previous setting on my code, and changing the TIME_ZONE of my RMariaDB Connection to SYSTEM was a bad idea, as it produced wrong results when running the procedures.

So the following is WRONG:
```r
dbExecute(
   con,
   "SET SESSION TIME_ZONE = SYSTEM;"
)
```
The good news is that when executing my conflictive procedure again (Which gave me wrong results in the past) either with:
* odbc (With MySQL ODBC 8.0 Connector).
* RMariaDB 1.0.6
* MariaDB IDE.

In all cases the results where the same (In a past time they weren't), so the conflictive setting seems to be fixed with RMariaDB 1.0.6 :) :) :)

Thank you,
Juan

**Comment by krlmlr** [comment]
@SymbolixAU: Thanks for the heads up, it seems that something's missing indeed:

``` r
library(DBI)

con <- dbConnect(RMariaDB::MariaDB(), dbname = "test")
dbWriteTable(con, "test", data.frame(time = Sys.time()), temporary = TRUE)
dbGetQuery(con, "SELECT NOW()")
#>                 NOW()
#> 1 2018-12-30 16:33:39
dbGetQuery(con, "SELECT * FROM test")
#>                  time
#> 1 2018-12-30 16:33:39

dbExecute(con, "SET time_zone = '-05:00'")
#> [1] 0
dbGetQuery(con, "SELECT NOW()")
#>                 NOW()
#> 1 2018-12-30 11:33:39
dbGetQuery(con, "SELECT * FROM test")
#>                  time
#> 1 2018-12-30 16:33:39
```

<sup>Created on 2018-12-30 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

Closing this in favor of # 116 (link above).
